pheps/phep-0001.md

```diff
@@ -0,0 +1,368 @@
+
+Once a PHEP has been accepted, its status will be changed to "Final".
+
+If no progress is being made on the PHEP, the editor should consult the author. The author may resume work on the PHEP, [transfer it to another author](transferring-phep-ownership), or withdraw it. If the author takes none of these actions, the editor may choose to bring the PHEP to the community for acceptance or rejection in its current form.
+
+If a PHEP fails to find consensus, the author may choose to modify it for reconsideration. The PHEP may be "Withdrawn" if the PHEP author themself decides that the PHEP is actually a bad idea, or has accepted that a competing proposal is a better alternative. A PHEP can also "Rejected" if it fails to find consensus and the author is not willing or able to modify it. Perhaps after all is said and done it was not a good idea. It is still important to have a record of this fact.
+
+When a PHEP changes state, the PHEP preamble must be updated accordingly. In addition to updating the Status field, the Resolution header must be added with a direct link to the minutes of the telecons and meetings where a decision on the PHEP was made. The editors must post announcements of PHEP resolution to the PyHC mailing list.
+
+Final PHEPs can also be replaced by a different PHEP, rendering the original obsolete.
+
+The possible paths of the status of PHEPs are as follows:
+
+![PHEP status flow diagram](phep-0001/process_flow.svg)
+
+Upon reaching "Final", "Withdrawn", or "Rejected" status, the editors update the preamble and merge the PR into `main` (maintaining the revision history). They tag the resulting commit and mint a DOI. If the new PHEP replaces an existing one, the new DOI should be marked as a new version of the previous, and the Replaced PHEP must have the `Replaced-By` field in its header updated.
+
+The complete PHEP approval process is:
+
+![PHEP workflow diagram](phep-0001/workflow.svg)
+
+The author may withdraw the PHEP at any time, so this is not shown. Replacement of a PHEP is not shown as it results from the approval of a new, replacement PHEP using the same process.
+
+## PHEP Maintenance
+<a name="phep-maintenance"></a>
+PHEPs are no longer substantially modified after they have left the Draft state. Once resolution is reached, a PHEP is considered a historical document rather than a living specification and the PHEP number is a permanent identifier for the document.
+
+Rejected and Withdrawn PHEPs must not be resurrected but may be used as the basis for a new PHEP.
+
+Changes to PHEPs in the Final state are limited to the preamble, and only in the event the PHEP is replaced. Final PHEPs are immutable and must be replaced by new PHEPs if necessary, rather than updated. The scope of a PHEP should be carefully considered in this light. Preamble changes are made via PR to the `standards` repository, merged by PHEP editors.
```

Comment:
A consequence of PHEPs being immutable is that making a minor change to a PHEP would require creating an entirely new PHEP. I'm hesitant about this since this could effectively end up being a change preventer. 

For example, there might be a time where we decide we want to change the process for approving a PHEP. If the change is fairly minor or incremental, it probably wouldn't feel worth the effort to create an entirely new PHEP and go through this process once again. If we were to create an entirely new PHEP for a minor change, we would also have a harder time looking at the version control history of the PHEP(s) which would make it more difficult to figure out what the changes actually were.

I'm wondering if there is a way that we could preserve immutability while making it easier to amend PHEPs. **What would we think of having PHEPs be versioned?**  That way, if we want to change the decision-making process, we could preserve the PHEP number and have separate DOIs for versions 1 & 2. The version number could be kept in the metadata for the PHEP. The mechanics for doing this are already in place on Zenodo, where authors can upload new versions of records.

(We also have a [section on amending PLEPs in PLEP 1](https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0001.rst#amending-or-superseding-a-plep) for PlasmaPy.)

Reply:
I think we're going to need more people to chime in on this one.

Reply:
I like the versioning idea Nick brings up here, but I hesitate about leaving such changes up to a committee. In some cases, changing one word in a PHEP can have large consequences. I also like the idea of recommending discussion with the original author of the PHEP being amended (if possible). How about allowing PHEPs to be amended, including the approval discussion/process on GitHub, but only requiring one vote once consensus is reached?

Reply:
We were discussing this earlier in another comment thread but it died down. @namurphy, I'd actually asked if you could provide an example or two of a time where you amended a PLEP and were happy you could do so, and it'd still be helpful for me to hear such an example. Because I (and I think Jon too) lean against the idea of versioning PHEPs, and hence having to refer to each one as **PHEP X version Y**. It just seems simpler to me to make them immutable. Although I recognize that means exercising effort to make sure you get a PHEP exactly right the first try.

Reply:
Sure!  One example for us is [PLEP 7](https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0007.rst) which lays out the structure of top-level subpackages in PlasmaPy. When we want to add a new subpackage, we go through the process of amending the PLEP. This has the effect that PLEP 7 is essentially a living document.  

In practice, I don't think I've ever had the need to refer to something as PLEP X version Y since we can usually implicitly refer to the most recent version. If we end up having to cite it, we can use the version number along with the DOI for that exact version. 

Overall, it's worth thinking about the tradeoffs between having PHEPs be immutable vs. versioned. Immutability will make PHEPs more stable and reduce possible ambiguity, while allowing PHEPs to be amended will allow for more flexibility and making it a lot easier to make minor changes.

Reply:
@namurphy , thanks for engaging with this...it's a tricky thing to try to resolve asynchronously and I appreciate the work. I've dug up the [previous conversation](https://github.com/heliophysicsPy/standards/pull/22#pullrequestreview-1695467345) so people can look at that. Every point you bring up is really making me think, even if I'm not changing my mind :)

I've got a few more thoughts, but I'm also wondering, would it make sense for me to try and summarize the differences between the approaches in a table or something? I'm thinking of clarifying our own thinking, but also making it easy for people who haven't engaged in this discussion to understand the trade-off and weigh in.

Reply:
I personally want to discourage having any PHEP being a “living document.” I do want more finality than that. But I buy the argument that, in practice, one can assume you implicitly mean the latest version of a PHEP—so we wouldn’t _always_ have to refer to them like PHEP x version y. Which _is_ a tick in the “pros” column.

This immutibility question is in my mind the last major issue to solve before we can call PHEP-1 done. 

I still find myself leaning towards keeping them immutable, but I gotta admit I don’t have strong reasons for why. Thinking out loud, but: would it be ridiculous to say PHEPS are immutable, but then if in the future we come across a use case where we really wanna edit a PHEP, I could introduce a new PHEP-9999 titled “JK, PHEPs can be edited”? 🤷‍♂️ Decisions are hard.

Reply:
Here's my attempt to summarize the differences in approach. @namurphy, please check to make sure I'm being fair to your proposal, happy to edit.

## Summary table for treating Final PHEPs as immutable vs. mutable.

For all TBD's, [PlasmaPy can be considered as an example](https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0001.rst#amending-or-superseding-a-plep).

||Immutable PHEPs|Mutable PHEPs|
|-|-|-|
| Process for amending | Copy existing PHEP, assign new number, include `Replaces` header, make edits, introduce as a new PHEP | TBD |
| Process for replacing | Copy existing PHEP, assign new number, include `Replaces` header, make edits, introduce as a new PHEP | Same (amendments may be sufficient, for many cases, but ability to replace is probably still needed for very large changes) |
| Approvals needed for changes | Same process as for any PHEP, two-round vote. | TBD, ideally "lighter", potentially single-round vote. |
| Distinguishing between "amendment" and "replacement" | N/A, any change requires a replacement | TBD, need to determine criteria for the decision and who makes it (e.g. original author, amendment author, PHEP editors, full vote |
| Determining latest version | Latest version will not have a `Replaced-By` header; otherwise use that header as reference to new version. DOI record of replaced PHEP should be updated to indicate DOI of replacement PHEP as well. | Latest version is that which is "live" on the PHEP site; new versions will also get a new DOI and similarly have updated records. Assume unqualified references to PHEP are to latest version unless specified. Also check `Replaced-by` header to see if fully replaced (instead of amended).|
| Checking against earlier version to see if something "used to be compliant" | Check the `Replaces` header to find PHEP number of earlier PHEPs, read on main site. Look at DOI record which should indicate "newer version of..." | Look at version number and find earlier version (potentially in git history if not on main site). Look at DOI record which should indicate "newer version of...". Also check `Replaces` header to see if fully replaced (instead of amended).|
| Deciding we made the wrong decision and going to the other approach | Introduce a new PHEP that replaces PHEP-1 and includes an amendment process. Difficult to retroactively declare certain PHEPs as "really amendments" and renumber, add version numbers, but maybe possible. | Introduce amendment or replacement to PHEP-1 that removes the amendment process. Include how previously-amended PHEPs will be handled (e.g. by retroactively assigning new PHEP numbers, or keeping the version numbers for those PHEPs but not future changes). |
| Implicit strategy | PHEPs should be general and flexible, allowing some change in implementation details as long as they remain compliant. Allows for flexibility "outside of the process". | PHEPs can be more specific and include more implementation details, as easier to amend. More flexibility "within the process", discourages flexibility "outside of the process" (i.e. more likely to include details that may change) |

Reply:
Here's where I abandon attempts at neutrality and fight my corner :)

- Easy amendment, move-fast seems more appropriate to me in the case of a tightly-coupled team, where people are generally aware of what potential changes are "in the air". It is also more appropriate where decision-making resides in relatively few people (e.g. a steering committee). In the case of PyHC, there are a lot of people, PyHC is usually not our first priority, and (as currently written in PHEP-1), decision-making is spread across pretty much everybody. Those aspects of PyHC might change, of course, but I suspect it will always be more spread than, say, PlasmaPy.
- I have been thinking of standards as what to do and why, not so much how. The fast-changing details can live in the reference implementation, which is not subject to approvals after the fact. That does lead to a little bit of ambiguity but I'm not sure every time some detail changes there needs to be approval--it just needs to be clear who's responsible for keeping things going, and within what limits. Even in this case I set up #24 separate from the idea of the process.
- Immutable PHEPs means changes have to go through a full process, which is in some sense more complicated. But it also means we have _only one process for approvals_, which is in some sense _less_ complicated. We wouldn't need to define amendment vs. replacement, what is different about the amendment process, etc. A "heavier" process, but one that meets multiple needs without having to introduce additional processes.
- I'm wary of a "streamlined" approval process that might allow for "back door" wholesale rewriting, sort of the [SPAC](https://en.wikipedia.org/wiki/Special-purpose_acquisition_company) of standards. A lot of pathologies that have come up in our discussion are addressed by the two-step vote. Re-examining that discussion in light of a lightweight amendment process is a challenge--which may still be worth the effort!
- To @sapols' point, I think it's probably easier to go from "immutable" to "mutable" than the other way around, although either direction will leave a bit of annoying cruft around.

I want to solicit more voices here. @rebeccaringuette you've chimed in, further thoughts based on my table above and the additional conversation?

Our other package reps (@aburrell @nabobalis @rweigel @jameswilburlewis), what are your thoughts on this discussion?

Reply:
An example of the immutable approach is RFCs, e.g. [RFC5322](https://datatracker.ietf.org/doc/html/rfc5322) and [RFC2549](https://datatracker.ietf.org/doc/html/rfc2549). I don't know the full RFC process, but it's interesting to see that those aren't even updated to show the new one (you can walk backwards in history, but not forward), which is _not_ what I propose here.

PEP1, incidentally, takes the mutable approach (without explicit version numbering).